controllers/complianceeventsapi/server.go

```diff
@@ -446,9 +503,145 @@ func parseQueryArgs(queryArgs url.Values, isCSV bool) (*queryOptions, error) {
+) (authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters []string,
+	includeAllManagedClusters bool, err error,
+) {
+	// Get all managedCluster rules
+	allRules, err := getManagedClusterRules(userConfig, nil)
+	if err != nil {
+		return authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters, false, err
+	}
+
+	clusterIDs, okClusterIds := parsed.Filters["clusters.cluster_id"]
+	if okClusterIds {
+		// Convert id to name
+		for _, id := range clusterIDs {
+			clusterName, err := getClusterNameFromID(ctx, db, id)
+			if err != nil {
+				log.Error(err, "Failed to get cluster name from cluster id :"+id)
+
+				// This leads to send forbidden error to the user
+				unAuthorizedClusters = append(unAuthorizedClusters, id)
+
+				continue
+			}
+
+			if getAccessByClusterName(allRules, clusterName) {
+				authorizedClusterIds = append(authorizedClusterIds, id)
+			} else {
+				unAuthorizedClusters = append(unAuthorizedClusters, id)
+			}
+		}
+	}
+
+	parsedClusterNames, okClusterNames := parsed.Filters["clusters.name"]
+	if okClusterNames {
+		for _, clusterName := range parsedClusterNames {
+			if getAccessByClusterName(allRules, clusterName) {
+				authorizedClusterNames = append(authorizedClusterNames, clusterName)
+			} else {
+				unAuthorizedClusters = append(unAuthorizedClusters, clusterName)
+			}
+		}
+	}
+
+	// There is no cluster.id or clusterName arg from the url query
+	// In other words, the user requests all
+	if !okClusterIds && !okClusterNames {
+		startRules, ok := allRules["*"]
+		if ok && slices.Contains(startRules, "get") || slices.Contains(startRules, "*") {
+			return authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters, true, nil
+		}
+
+		managedClusterNames, err := getAllManagedClusterNames(ctx, c)
+		if err != nil {
+			return authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters, false, err
+		}
+
+		for _, mcName := range managedClusterNames {
+			// Should have "GET" auth for all managedCluster
+			if !getAccessByClusterName(allRules, mcName) {
+				unAuthorizedClusters = append(unAuthorizedClusters, mcName)
+			}
+		}
+
+		if len(unAuthorizedClusters) == 0 {
+			// includeAllManagedClusters is true
+			return authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters, true, nil
+		}
+	}
+
+	return authorizedClusterIds, authorizedClusterNames, unAuthorizedClusters, false, nil
+}
+
+func getAccessByClusterName(allManagedClusterRules map[string][]string, clusterName string) bool {
+	// User requested all managed cluster and has all managedCluster Auth
+	startRules, ok := allManagedClusterRules["*"]
+	if ok && slices.Contains(startRules, "get") || slices.Contains(startRules, "*") {
+		return true
+	}
+
+	rules, ok := allManagedClusterRules[clusterName]
+	if ok && slices.Contains(rules, "get") || slices.Contains(rules, "*") {
+		return true
+	}
+
+	return false
+}
+
+func getAllManagedClusterNames(ctx context.Context, c client.Client) ([]string, error) {
```

Comment:
After following this comment https://github.com/open-cluster-management-io/governance-policy-propagator/pull/164/files#r1491133119, you'll be able to remove this function.